src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -22,30 +23,45 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
     internal class SemanticTokensHelpers
     {
         /// <summary>
-        /// Returns the semantic tokens data for a given document with an optional range.
+        /// Returns the semantic tokens data for a given document with an optional ranges.
         /// </summary>
         public static async Task<int[]> ComputeSemanticTokensDataAsync(
             ClientCapabilities capabilities,
             Document document,
-            LSP.Range? range,
+            LSP.Range[]? ranges,
             ClassificationOptions options,
             CancellationToken cancellationToken)
         {
-            var tokenTypesToIndex = SemanticTokensSchema.GetSchema(capabilities.HasVisualStudioLspCapability()).TokenTypeToIndex;
+            if (ranges is not null)
+            {
+                // Expecting sorted ranges within the document
+                Array.Sort(ranges, static (r1, r2) => r1.CompareTo(r2));
+            }
 
+            var tokenTypesToIndex = SemanticTokensSchema.GetSchema(capabilities.HasVisualStudioLspCapability()).TokenTypeToIndex;
             var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
             var text = await document.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+            var classifiedSpansLength = ranges?.Length ?? 1;
+            using var _ = ArrayBuilder<ClassifiedSpan>.GetInstance(classifiedSpansLength, out var classifiedSpans);
 
-            // By default we calculate the tokens for the full document span, although the user 
-            // can pass in a range if they wish.
-            var textSpan = range is null ? root.FullSpan : ProtocolConversions.RangeToTextSpan(range, text);
+            for (var i = 0; i < classifiedSpansLength; i++)
```

Comment:
>for (var i = 0; i < classifiedSpansLength; i++) [](http://example.com/codeflow?start=12&length=47)

nit: if you use the suggestion above to ensure ranges isn't null, this can be a foreach